SU2_CFD/include/variables/CTurbVariable.hpp

```diff
@@ -27,22 +27,19 @@
 
 #pragma once
 
-#include "CVariable.hpp"
+#include "CScalarVariable.hpp"
 
 /*!
  * \class CTurbVariable
  * \brief Base class for defining the variables of the turbulence model.
  * \ingroup Turbulence_Model
  * \author A. Bueno.
  */
-class CTurbVariable : public CVariable {
+class CTurbVariable : public CScalarVariable {
 protected:
   VectorType muT;         /*!< \brief Eddy viscosity. */
```

Comment:
The solvers see the nodes as CScalarVariable, but the methods to access muT are in CTurbVariable.
The compiler does not know that the type of nodes will be CTurbVariable at runtime and so it cannot de-virtualize the access to variables that are not in CScalarVariable.

By making CScalarSolver a class template you can instantiate it with different types for "nodes", i.e. in practice you end up with multiple CScalarSolver "tuned" for each of the derived classes. CScalarSolver<CTurbVariable>, CScalarSolver<CHeatVariable> (which are not the same solver anymore, but that is fine because we want this as a building block.

Reply:
I did a little debugging on restarts over at feature_flamelet and feature_multicomp, so this will be a bit delayed 🐌 sorry for that

Reply:
Should I go for a similar approach as `CFVMFlowSolverBase.hpp/inl` with the explicit instantiation of the Solver i need in the derived solvers? i.e. `template class CScalarSolver<CTurbVariable>;` at the top of CTurbVariable.cpp?

alternatively I could include CScalarSolver.cpp (or rename to .inl) at the end of its hpp?

Reply:
Do it like CFVMFlowSolverBase.hpp and .inl please.

Reply:
Implemented in ca16276d50b2dc99ea40c511aa6df70b141b986c Let me know if that is what you think should be done

Reply:
Just for my own documentation: [template use in implementation file](https://stackoverflow.com/questions/18186878/invalid-use-of-template-name-without-an-argument-list) / [templates implemented in header files](https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file)